File: 03_Matrices/c/matrix.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "matrix.h"

/*-------------- struct matrix *new_matrix() --------------
Inputs:  int rows
         int cols 
Returns: 

Once allocated, access the matrix as follows:
m->m[r][c]=something;
if (m->lastcol)... 
*/
struct matrix *new_matrix(int rows, int cols) {
  double **tmp;
  int i;
  struct matrix *m;

  tmp = (double **)malloc(rows * sizeof(double *));
  for (i=0;i<rows;i++) {
      tmp[i]=(double *)malloc(cols * sizeof(double));
    }

  m=(struct matrix *)malloc(sizeof(struct matrix));
  m->m=tmp;
  m->rows = rows;
  m->cols = cols;
  m->lastcol = 0;

  return m;
}


/*-------------- void free_matrix() --------------
Inputs:  struct matrix *m 
Returns: 

1. free individual rows
2. free array holding row pointers
3. free actual matrix
*/
void free_matrix(struct matrix *m) {

  int i;
  for (i=0;i<m->rows;i++) {
      free(m->m[i]);
    }
  free(m->m);
  free(m);
}


/*======== void grow_matrix() ==========
Inputs:  struct matrix *m
         int newcols 
Returns: 

Reallocates the memory for m->m such that it now has
newcols number of collumns
====================*/
void grow_matrix(struct matrix *m, int newcols) {
  int i;
  for (i=0;i<m->rows;i++) {
      m->m[i] = realloc(m->m[i],newcols*sizeof(double));
    }
  m->cols = newcols;
}

/*-------------- void copy_matrix() --------------
Inputs:  struct matrix *a
         struct matrix *b 
Returns: 

copy matrix a to matrix b
*/
void copy_matrix(struct matrix *a, struct matrix *b) {
  int r, c;

  for (r=0; r < a->rows; r++) 
    for (c=0; c < a->cols; c++)  
      b->m[r][c] = a->m[r][c];  
}




/*-------------- void print_matrix() --------------
Inputs:  struct matrix *m 
Returns: 

print the matrix
*/
void print_matrix(struct matrix *m) {
  int i, j;
  for(i = 0; i < (*m).cols; i++){
    for(j = 0; j < (*m).rows; j++)
      printf("%lf ", (*m).m[i][j]);
    printf("\n");
  }
  printf("\n");
}

/*-------------- void ident() --------------
Inputs:  struct matrix *m <-- assumes m is a square matrix
Returns: 

turns m in to an identity matrix
*/
void ident(struct matrix *m)
{
  int x,y;
  for(x = 0; x < m->cols; x++)
  {
    for(y = 0; y < m->rows; y++)
    {
      m->m[x][y] = (x == y);
    }
  }
}


/*-------------- void scalar_mult() --------------
Inputs:  double n
         struct matrix *m 
Returns: 

multiply each element of m by n
*/
void scalar_mult(double n, struct matrix *m)
{
  int x,y;
  for(x = 0; x < m->cols; x++)
  {
    for(y = 0; y < m->rows; y++)
    {
      m->m[x][y] *= n;
    }
  }
}


/*-------------- void matrix_mult() --------------
Inputs:  struct matrix *a
         struct matrix *b 
	 struct matrix *m
Returns: 

*/
void matrix_mult(struct matrix *a, struct matrix *b, struct matrix *m)
{
  int x,y;
  int i;
  for(x = 0; x < m->cols; x++)
  {
    for(y = 0; y < m->rows; y++)
    {
      m->m[x][y] = 0;
      for(i = 0; i < a->cols && i < b->rows; i++)
      {
	m->m[x][y] += ((a->m[x][i]) * (b->m[i][y]));
      }
    }
  }
}

int main()
{
  struct matrix* mat = new_matrix(4, 4);
  struct matrix* m2 = new_matrix(4, 4);
  struct matrix* m3 = new_matrix(4, 4);
  int i, j;
  print_matrix(mat);
  ident(mat);
  print_matrix(mat);
  mat->m[1][2]=5;
  mat->m[1][3]=7;
  print_matrix(mat);
  scalar_mult(2, mat);
  print_matrix(mat);
  scalar_mult(-3, mat);
  print_matrix(mat);
  ident(m2);
  matrix_mult(mat, m2, m3);
  print_matrix(m3);
  matrix_mult(m2, mat, m3);
  print_matrix(m3);
  for(i = 0; i < 4; i++)
  {
    for(j = 0; j < 4; j++)
    {
      m2->m[i][j] = i * j + 3 + j * j + i;
    }
  }
  print_matrix(m2);
  matrix_mult(m3, m2, mat);
  print_matrix(mat);
  return 1;
}

File: 03_Matrices/java/Matrix.java
/*========== Matrix.java ==========
  Matrix will hold a 2-d array of doubles and have a default size of 4x4.
  Handles basic matrix maintenence and math.
  Creates transformation matricies for tralation, scale and rotate
=========================*/

import java.io.*;
import java.util.*;

public class Matrix {

    public static final int DEFAULT_SIZE = 4;
    protected double[][] m;

    /*===========Constructors================
      Default constructor creates a 4x4 matrix
      Second constructor creates a 4xN matrix
    */
    public Matrix()
    {
	m = new double[DEFAULT_SIZE][DEFAULT_SIZE];
    }
    
    public Matrix(int c)
    {
	m = new double[DEFAULT_SIZE][c];
    }
    
    public Matrix(int r, int c)
    {
	m = new double[r][c];
    }
    
    /*===========grow================
      Increase the number of columns in a matrix by 10
      You can change the growth factor as you see fit
    */
    public void grow()
    {
	double[][] n = new double[m.length][m[0].length + 10];
	for (int r = 0; r < m.length; r++)
	{
	    for (int c = 0; c < m[r].length; c++)
	    {
		n[r][c] = m[r][c];
	    }
	}
	m = n;
    }

    /*======== public void clear() ==========
      Inputs:  
      Returns: 
      Sets every entry in the matrix to 0
      ====================*/
    public void clear()
    {
	for (int i = 0; i < m.length; i++)
	{
	    for (int j = 0; j < m[i].length; j++)
	    {
		m[i][j] = 0;
	    }
	}
    }

    /*===========ident================
      Turns this matrix into the indentity matrix
      You may assume the calling Matrix is square
    */
    public void ident()
    {
	for (int i = 0; i < m.length; i++)
	{
	    for (int j = 0; j < m[i].length; j++)
	    {
		m[i][j] = i == j ? 1 : 0;
	    }
	}
    }

    /*===========scalarMult================
      Inputs:  double x
      
      multiply each element of the calling matrix by x
    */
    public void scalarMult(int s)
    {
	for (int i = 0; i < m.length; i++)
	{
	    for (int j = 0; j < m[i].length; j++)
	    {
		m[i][j] = s * m[i][j];
	    }
	}
    }		

    /*===========matrixMult================
      Multply matrix n by the calling matrix, modify
      the calling matrix to store the result.
      
      eg.
      In the call a.matrixMult(n), n will remain the same
      and a will now be the product of n * a
    */
    public void matrixMult(Matrix n)
    {
	double[][] tmp = new double[n.m.length][m[0].length];
	for(int r = 0; r < tmp.length; r++)
	{
	    for(int c = 0; c < tmp[r].length; c++)
	    {
		tmp[r][c] = 0;
		for(int i = 0; i < m.length && i < n.m.length; i++)
		{
		    tmp[r][c] += n.m[r][i] * m[i][c];
		}
	    }
	}
	m = tmp;
    }
   
    /*===========copy================
      Create and return new matrix that is a duplicate 
      of the calling matrix
    */
    public Matrix copy()
    {
	Matrix result = new Matrix(m.length, m[0].length);
	for(int r = 0; r < m.length; r++)
	{
	    for(int c = 0; c < m[r].length; c++)
	    {
		result.m[r][c] = m[r][c];
	    }
	}
	return result;
    }

    /*===========toString================
      Crate a readable String representation of the 
      calling matrix.
    */
    public String toString()
    {
	String result = "";
	for (int i = 0; i < m.length; i++)
	{
	    for (int j = 0; j < m[i].length; j++)
	    {
		result = result + m[i][j] + " ";
	    }
	    result = result + "\n";
	}
	return result;
    }

}
